apps/src/lib/node/ledger/protocol/transactions/validator_set_update/mod.rs

```diff
@@ -31,29 +34,116 @@ where
     D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
     H: 'static + StorageHasher + Sync,
 {
+    if ext.signatures.is_empty() {
+        tracing::debug!("Ignoring empty validator set update");
+        return Ok(Default::default());
+    }
+
     tracing::info!(
         num_votes = ext.signatures.len(),
         "Aggregating new votes for validator set update"
     );
 
     let voting_powers = utils::get_voting_powers(storage, &ext)?;
-    let changed_keys = apply_updates(storage, ext, voting_powers)?;
+    let changed_keys = apply_update(storage, ext, voting_powers)?;
 
     Ok(TxResult {
         changed_keys,
         ..Default::default()
     })
 }
 
-fn apply_updates<D, H>(
-    _storage: &mut Storage<D, H>,
-    _ext: validator_set_update::VextDigest,
-    _voting_powers: HashMap<(Address, BlockHeight), FractionalVotingPower>,
+fn apply_update<D, H>(
+    storage: &mut Storage<D, H>,
+    ext: validator_set_update::VextDigest,
+    voting_powers: HashMap<(Address, BlockHeight), FractionalVotingPower>,
 ) -> Result<ChangedKeys>
 where
     D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
     H: 'static + StorageHasher + Sync,
 {
-    tracing::warn!("Called apply_updates() with no side effects");
-    Ok(ChangedKeys::new())
+    let epoch = {
+        // all votes we gathered are for the same epoch, so
+        // we can just fetch the block height from the first
+        // signature we iterate over, and calculate its cor-
+        // responding epoch
+        let height = ext
+            .signatures
+            .keys()
+            .map(|(_, height)| *height)
+            .by_ref()
+            .next()
+            .expect(
+                "We have at least one signature present in this validator set \
+                 update vote extension digest",
+            );
+
+        storage
+            .get_epoch(height)
+            .expect("The epoch of the given block height should be known")
+    };
+
+    let valset_upd_keys = vote_tallies::Keys::from(&epoch);
+    let (exists_in_storage, _) = storage.has_key(&valset_upd_keys.seen())?;
+
+    let mut seen_by = Votes::default();
+    for (address, block_height) in ext.signatures.into_keys() {
+        // TODO: more deterministic deduplication
```

Comment:
what's wrong we non-deterministic de-duplication?

Reply:
this arbitrarily picks the first vote we iterate over in the `HashMap` of signatures. we could pick, e.g. the vote at the earliest block height instead.

Reply:
I will update this to use the earliest one by block beight (https://github.com/anoma/namada/pull/517) (sorry should have linked the TODO here to an issue)